L64-python-runtime-proto-parsing.md

```diff
@@ -0,0 +1,99 @@
+
+At the least, this results in surprises when an update to a `.proto` file does not result in an update to the behavior of the user’s code. However, when client and server code lives in separate repos, this can result in aliasing, where one repository houses generated code from an earlier version of the protocol buffer than the other.
+
+Open source users are aware of this gap in the ecosystem and are actively looking for ways to fill it. Many have settled on [protocol buffer monorepos](https://medium.com/namely-labs/how-we-build-grpc-services-at-namely-52a3ae9e7c35) as a solution to the problem, wherein all `.proto` files for an organization are placed in a single source code repository and included by all other repositories as a submodule. But even this is not a complete solution. In addition, some mechanism must be put in place for the repositories housing the client and server code to retrieve the desired protocol buffer and generate code for the target language.
+
+The protocol compiler paired with Google's build system also means that the average engineer never has to manually invoke `protoc`. Instead, when an update is made to a `.proto` file and some code file references it, the code for the protocol buffer is regenerated without any manual intervention on the part of the engineer. Compare that to today's workflow for gRPC users:
+
+ 1. Update .proto file.
+ 2. Manually regenerate code. (remembering how to use all of the CLI flags)
+ 3. Make necessary corresponding updates to code using the protocol buffer.
+ 4. Rerun.
+
+It's easy for several of those steps to slip one's mind while developing. Moreover, figuring out how to invoke `protoc` in a way that meshes with your imports can be quite difficult. Python developers in particular are unused to build-time steps such as these; it is much more common to perform them at runtime.
+
+### Related Proposals
+
+* [Node.js gRPC+Protobuf.js Library](https://github.com/grpc/proposal/blob/master/L23-node-protobufjs-library.md)
+
+
+## Proposal
+
+Whereas today, users import protobuf code by invoking `protoc` to generate a file called `foo_pb2.py` and including the following line of code:
+
+```python
+import foo_pb2
+import foo_pb2_grpc
+```
+
+They will now also have the option of completely skipping the manual protoc invocation and instead writing
+
+```python
+protos = grpc.protos('foo.proto')
```

Comment:
Would be any chance on having a more explicit API for returning explicitly a specific module, for example:

```python
foo_pb2 = grpc.protos('foo.proto',  modules=["foo_pb2"])
```
This will make the API more explicit, which would help users and developers to not have the cognitive burden of remembering that modules were magically imported behind the scenes.

Another concern of not having an explicit creation of the module would be with some complaining that syntax checkers might give you, which they would be false positive. Have you checked what would happen with pylint or IDEs like PyCharm?

Reply:
@pfreixes Thanks for the suggestion!

The problem is that whenever someone specified a module name besides the default (`foo_pb2` in your example), they'd introduce the possibility of the unintentionally loading a logically equivalent, but nevertheless incompatible module into the process. For example:

```python
foo = grpc.protos('foo.proto', modules=["foo"]) # load from .proto file
import foo_pb2 # load from pre-generated code by accident
```

Types from `foo` and `foo_pb2` look the same. They have the same names, constructors, and methods. But they are completely incompatible. Even if they hold the same values, they will never evaluate as equal. Containers from `foo` cannot be filled with elements from `foo_pb2` and vice versa.

The usage of the `foo_pb2` format as the canonical name is meant be a detail that can just be forgotten about, even in the case that the process mixes pre-generated code and runtime parsing.

You make an interesting point about syntax checkers. But if they do have a problem, it will exist for anything that uses `importlib`, including modules loaded from zip files, modules loaded from shared object libraries, and modules loaded from `.pyc` files with no corresponding `.py` file.

Reply:
Regarding the example that you gave, is there a typo? Should be: 

```python
foo = grpc.protos('foo.proto', modules=["foo_pb2"]) # load from .proto file
import foo_pb2 # load from pre-generated code by accident
```
?

If so, why the user might end up mixing both implementations? just because different modules, or even packages, might be loading objects in a different way? for example, the caller which might be the main module using the new `grpc.protos` and the dependency might be still using the `import foo_pb2`?

Why the implementations would be different? I thought - maybe I'm totally wrong - that the `protos` field was indeed generating in memory the same python file that would be generated in disk. If so, why they would need to be considered incompatible?

Reply:
There was no typo in my example. I explicitly chose a module name besides what would be the default for such an option. The example that you included above is equivalent to the following in the current proposal.

```python
foo = grpc.protos('foo.proto') # Import from a ".proto" file
import foo_pb2 # Does not trigger a re-import.
assert foo is foo_pb2
```

In my example, however,
```python
foo = grpc.protos('foo.proto', modules=["foo"])
import foo_pb2
foo.FooMessage(a=2) == foo_pb2.FooMessage(a=2)  # This will always be false.
foo.FooMessage is foo_pb2.FooMessage  # Because this is always false.
```
> If so, why the user might end up mixing both implementations? just because different modules, or even packages, might be loading objects in a different way? for example, the caller which might be the main module using the new grpc.protos and the dependency might be still using the import foo_pb2?
> Why the implementations would be different? I thought - maybe I'm totally wrong - that the protos function was indeed generating in memory the same python file that would be generated in disk. If so, why they would need to be considered incompatible?

The problem isn't that there are two modules loaded in different ways. The problem is that there are two modules *at all*. This comes down to two things:

- [How imports in a proto file are handled](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/compiler/python/python_generator.cc#L430) (If you didn't put it under its canonical name, the import mechanism within the generated source code won't be able to find it)
- [The protobuf C extension requires pointer equality of the descriptors backing message types for them to work.](https://github.com/protocolbuffers/protobuf/blob/3be9322e28dfcacdb669f646f76fe833fa274107/python/google/protobuf/pyext/message.cc#L2176). This effectively kills interoperability through duck typing.

The only way to guarantee compatibility is to ensure that there is only ever one module loaded in the interpreter for each Protobuf file, regardless of whether the code was generated at build time or at runtime.

Reply:
Oks got it, it seems that the main issue for having compatibility comes from the fact that the comparison would not work since pointers are being compared. But in any case, I would kinda agree that having two different modules loaded and trying to make them compatible could become a nightmare.

Sorry for the usage of the `modules` parameter, in somehow I wanted to add something for narrow down the modules and I ended up mixing two different things: modules and messages.

Oks let me just add a last little detail, should we prescript the user for doing something like this:

```python
foo_pb2 = grpc.protos('foo.proto')
```

Again trying to be as much explicit as we can, providing this recipe  the user would make something like:

```python
foo_pb2 = grpc.protos('foo.proto')

def create_message(a: int) -> foo_pb2.FooMessage:
    return foo_pb2.FooMessage(a=a)
```

Reply:
So, I explicitly chose not to include the `_pb2` suffix because it's confusing. Newcomers to gRPC have no idea what it is. When I write a file called "foo.py", I expect to be able to `import foo`, not `import foo_py3`.

You follow up with an example about typing. I don't think that the example you included would actually work. I'm not super familiar with how Python type checkers work, but I'm fairly certain they don't execute the full source code of the file being checked. In general though, I do see good integration with type checkers as being valuable. I'll do some investigation into how the feature as implemented interacts with them.

Reply:
@pfreixes I dug into the type checking a bit more. It turns out that this is just [not supported by the current Python type checkers](https://github.com/python/mypy/issues/5059). From the issue,
 
> `importlib.import_module` returns type `types.ModuleType`. We don't know anything else about the attributes of that module, beyond the attributes of that type. Therefore, the error message boils down to #1424.

> Perhaps we should warn when someone calls __import__, importlib.import_module or the like, because it almost always requires a type ignore.

At least for the moment, I would recommend that people wanting support for type checking continue to statically generate code and pull that in.

Reply:
Thanks for having a look, just for clarifying would this be fixed by doing 
```python
protos = grpc.protos('foo.proto')
import foo_pb2
```
So being foo_pb2 a noop but, maybe, allowing the type checker to know that there is behind?

Out of curiosity, would this issue happen also with `pytype` ?

Reply:
I gave the `import foo_pb2` approach a shot as well, but alas, it doesn't work either.

`pytype` behaves slightly differently, but also doesn't satisfactorily work. Instead of throwing an error about an unknown module, it seems to simply ignore types dependent on the unknown module. For instance, you could change your example above to the following without an error:

```python
foo_pb2 = grpc.protos('foo.proto')

def create_message(a: int) -> foo_pb2.NonexistentType:
    return foo_pb2.FooMessage(a=a)
```

Reply:
sad panda.

We should add a disclaimer since typing is nowadays well spread and used for many and many projects. And the side effect If I'm not wrong when mypy is used is an annoying exception.